src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServer.cs

```diff
@@ -213,28 +215,41 @@ public void Start<TContext>(IHttpApplication<TContext> application)
             }
         }
 
-        public void Dispose()
+        // Graceful shutdown if possible
+        public async Task StopAsync(CancellationToken cancellationToken)
         {
+            _isStopping = true;
             if (_transports != null)
             {
                 var tasks = new Task[_transports.Count];
                 for (int i = 0; i < _transports.Count; i++)
                 {
                     tasks[i] = _transports[i].UnbindAsync();
                 }
-                Task.WaitAll(tasks);
+                await Task.WhenAll(tasks);
 
                 // TODO: Do transport-agnostic connection management/shutdown.
                 for (int i = 0; i < _transports.Count; i++)
                 {
                     tasks[i] = _transports[i].StopAsync();
                 }
-                Task.WaitAll(tasks);
+                await Task.WhenAll(tasks);
             }
 
             _dateHeaderValueManager?.Dispose();
         }
 
+        // Ungraceful shutdown
+        public void Dispose()
+        {
+            if (!_isStopping)
```

Comment:
I would have StopAsync guard itself being called twice instead of checking it here. Just call StopAsync every time.